javalib/src/main/scala/java/lang/Long.scala

```diff
@@ -226,9 +226,101 @@ object Long {
     approxQuot.toString(radix) + paddingZeros.jsSubstring(remStr.length) + remStr
   }
 
+  private def parseLongFail(s: String): Nothing =
+    throw new NumberFormatException(s"""For input string: "$s"""")
+
+  @inline def parseLong(s: String): scala.Long =
+    parseLongPlatform(s, 10)
+
+  @inline // because radix is almost certainly constant at call site
   def parseLong(s: String, radix: Int): scala.Long = {
+    if (radix < Character.MIN_RADIX || radix > Character.MAX_RADIX)
+      parseLongFail(s)
+    parseLongPlatform(s, radix)
+  }
+
+  // Must be called only with a valid radix
+  @inline def parseLongPlatform(s: String, radix: Int): scala.Long = {
+    if (LinkingInfo.isWebAssembly)
+      parseLongImplWasm(s, radix, divideUnsigned(Int.MinValue, radix.toLong))
+    else
+      parseLongImplJS(s, radix)
+  }
+
+  /* Must be called only with a valid radix.
+   *
+   * The overflowBarrier must be divideUnsigned(Long.MinValue, radix.toLong).
+   *
+   * See Integer.parseIntImpl for the reasoning.
+   */
+  @noinline
+  private def parseLongImplWasm(s: String, radix: Int,
+      overflowBarrier: scala.Long): scala.Long = {
+
+    // !!! Begin duplicate code with Integer.parseIntImpl !!!
+
+    def fail(): Nothing = parseLongFail(s)
+
+    // Early checks: s non-null and non-empty
+    if (s == null)
+      fail()
+    val len = s.length
+    if (len == 0)
+      fail()
+
+    // Load the module instance of Character once, instead of in every loop iteration
+    val character = Character
+
+    /* Process the sign character.
+     * Set `sign` to `-1` if there is a leading '-', and `0` otherwise.
+     * Set `i` to 1 if there was a leading '+' or '-', and 0 otherwise.
+     */
+    val firstChar = s.charAt(0)
+    val negative = firstChar == '-'
+    val sign = if (negative) -1L else 0L
+    var i = if (negative || firstChar == '+') 1 else 0
+
+    // We need at least one digit
+    if (i >= len)
+      fail()
+
+    var result: scala.Long = 0L
+
+    while (i != len) {
+      val digit = character.digitWithValidRadix(s.charAt(i), radix).toLong
+      if (digit == -1L || (result ^ SignBit) > (overflowBarrier ^ SignBit))
+        fail()
+      result = result * radix + digit
+      /* The above addition can overflow the range of valid results (but it
+       * cannot overflow the unsigned int range). If that happens during the
+       * last iteration, we catch it with the final overflow check. If it
+       * happens during an earlier iteration, we catch it with the
+       * `overflowBarrier`-based check.
+       */
+      i += 1
+    }
+
+    /* Final overflow check. So far we computed `result` as an unsigned
+     * quantity. If negative, the maximum unsigned value allowed in
```

Comment:
```suggestion
     * quantity. If negative, the maximum unsigned value allowed is
```

Also in the original code in Integer.